### ISSUE TITLE
Apple's Contacts Support

### ISSUE THREAD

**Issue by q6276270** [Kozea/Radicale#740]
Hello,
is there any possibility for support of Apple’s Contacts on Mac OSX? I’ve tried to login on both on radicale 0.9 and 2.x and both times (for 2.x considering the fix given in #269) I managed to login but not to write to contacts.
Thanks for your time!

**Comment by martinhoefling** [comment]
I have the same problem, radicale 2.0.8, here's the log of my reverse proxy.

```
- [08/Nov/2017:20:39:37 +0000] "PROPFIND /.well-known/carddav HTTP/1.1" 404 60 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
- [08/Nov/2017:20:39:37 +0000] "PROPFIND / HTTP/1.1" 401 61 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
martin [08/Nov/2017:20:39:37 +0000] "PROPFIND / HTTP/1.1" 207 230 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
martin [08/Nov/2017:20:39:37 +0000] "OPTIONS /martin/ HTTP/1.1" 200 0 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
martin [08/Nov/2017:20:39:42 +0000] "PROPFIND /martin/ HTTP/1.1" 207 384 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
martin [08/Nov/2017:20:39:43 +0000] "PROPFIND /martin/ HTTP/1.1" 207 756 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
martin [08/Nov/2017:20:39:43 +0000] "PROPFIND /.well-known/carddav HTTP/1.1" 404 60 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
martin [08/Nov/2017:20:39:43 +0000] "PROPFIND / HTTP/1.1" 207 230 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
martin [08/Nov/2017:20:39:43 +0000] "OPTIONS /martin/ HTTP/1.1" 200 0 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
martin [08/Nov/2017:20:39:48 +0000] "PROPFIND /martin/ HTTP/1.1" 207 384 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
martin [08/Nov/2017:20:39:48 +0000] "PROPFIND /martin/ HTTP/1.1" 207 756 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
- [08/Nov/2017:20:46:29 +0000] "PROPFIND /martin/f4cda0ec-2f65-8f51-b17f-25bae7eed16e/ HTTP/1.1" 401 61 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
martin [08/Nov/2017:20:46:29 +0000] "PROPFIND /martin/f4cda0ec-2f65-8f51-b17f-25bae7eed16e/ HTTP/1.1" 207 298 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
- [08/Nov/2017:20:46:29 +0000] "OPTIONS /martin/ HTTP/1.1" 200 0 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
- [08/Nov/2017:20:46:34 +0000] "PROPFIND /martin/ HTTP/1.1" 401 61 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
martin [08/Nov/2017:20:46:34 +0000] "PROPFIND /martin/ HTTP/1.1" 207 384 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
martin [08/Nov/2017:20:46:34 +0000] "PROPFIND /martin/ HTTP/1.1" 207 756 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
martin [08/Nov/2017:20:46:35 +0000] "PROPFIND /martin/f4cda0ec-2f65-8f51-b17f-25bae7eed16e/ HTTP/1.1" 207 298 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
martin [08/Nov/2017:20:46:35 +0000] "OPTIONS /martin/ HTTP/1.1" 200 0 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
martin [08/Nov/2017:20:46:40 +0000] "PROPFIND /martin/ HTTP/1.1" 207 384 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
martin [08/Nov/2017:20:46:40 +0000] "PROPFIND /martin/ HTTP/1.1" 207 756 "-" "Mac OS X/10.13.1 (17B48) AddressBook/1806" "-"
```

After setting up, the account is shown in radicale but no contacts are shown. f4cda0ec-2f65-8f51-b17f-25bae7eed16e is an addressbook created and synced from another application.

Any ideas what could fix the problem?

**Comment by tmst** [comment]
I had Radicale 0.7 events and contacts set up on my old Mac running 10.5. Everything worked as expected.

**Comment by martinhoefling** [comment]
I tried Radicale 2.1.8 with high sierra contacts app.

**Comment by tmst** [comment]
Have you tried getting Radicale to work with any other client besides
the contacts app?
On Sat, Nov 11, 2017, at 02:05 PM, Martin Hoefling wrote:
> I tried Radicale 2.1.8 with high sierra contacts app.


> — You are receiving this because you commented. Reply to this email
> directly, view it on GitHub[1], or mute the thread[2].> 


Links:

  1. https://github.com/Kozea/Radicale/issues/740#issuecomment-343697314

**Comment by 1ecstasy** [comment]
No I haven't. I wanted my server software to be open source and to be able to interface with the native mac programs. Contacts are not that important to me. However I was hopeful that the issue would be resolved after some time. Mind you, I'm not trying to insist in any way. If support for Apple's contacts is not desired or too much work for whatever reason, just let me know and I'll be happy.

**Comment by return42** [comment]
Here, Radicale 2.x  on a Linux host behind [Apache](http://return42.github.io/handsOn/radicale/radicale_server.html) works like a charm with Apple's [Contacts](http://return42.github.io/handsOn/radicale/dav_client.html#apple-kontakte) and "Calendar" client. If you really want to get some help from community, please study the logs. If this does not help, come back with more detail information.

**Comment by Unrud** [comment]
> I have the same problem, radicale 2.0.8, here's the log of my reverse proxy.

We need the logging output from Radicale. Start it with the ``--debug`` argument.

> works like a charm with Apple's Contacts 

Does it work without any special configuration?

**Comment by HaJoSa** [comment]
I also tried Radicale 2.1.8 with high sierra contacts app and activated debugging.
But I can't log in to radicale: Contacts app throws an error - Account/Password could not be verified
(in German "Accountname/Passwort konnte nicht überprüft werden")
But the contactas app does not ask radicale at all, since there is not entry in the log file.
Btw. The calendar app work as expected with radicale on Mac High Sierra.

**Comment by HaJoSa** [comment]
After doing some research on the owncloud forum I found following solution
thanks to user calmch on central.owncloud.org:
- In the Add Account window I choose the "Manual" option;
- I entered with username and password;
- In the server address I entered with a complete address in one line, i. e.,
http://EXAMPLE.COM:MY_SERVER_PORT/USERNAME/ADdRESSBOOK

It seems that there is a bug in MacOS contacts client software, as it cannot compose the correct server link from the separated parameters we give to it. When we write manually the complete server link, as shown above, there is no need to compose the link and the login process succeeds.
I hope this solution could be useful to someone.

**Comment by 1ecstasy** [comment]
@HaJoSa 
I tried this some time ago. It was not successful, i.e. I couldn't login or I still couldn't write to the server.

**Comment by return42** [comment]
@HaJoSa : I guess you are german speaker .. may it helps if you read my (german) article about macOS Contacts & Radicale. http://return42.github.io/handsOn/radicale/dav_client.html#apple-kontakte

**Comment by HaJoSa** [comment]
@return42 : Thanks for the link.
Before my friend upgraded his Mac to High Sierra I never had problems with the contacts app.

**Comment by return42** [comment]
@HaJoSa : Does the link help?

> Before my friend upgraded his Mac to High Sierra I never had problems with the contacts app.

Ok, I have to say, that I have not tested with "high Sierra". What I have tested was "Sierra" (10.12.6). I have a handfull of old mac's with a remaining period. I don't buy these new macs (IMO crap hardware) and I upgrade mac OS as late as possible. My experience; every time I upgrade mac OS I got new iTunes & Cloud crap and apps like Calender & Contatcs are broken by some means or other .. Mac has had its days .. so sad.

**Comment by HaJoSa** [comment]
@return42 : your link is very helpfull.
I do not own a Mac myself but I give help to people with Macs ;-)
And they want support even on new system updates.

**Comment by return42** [comment]
> but I give help to people with Macs ;-) And they want support even on new system updates.

you got my commiseration :)  

If you think I can help you, contact me directly via eMail (my adress is given under the linked page) or leave an issue on https://github.com/return42/handsOn (its all german)

**Comment by oakbrad** [comment]
Thanks @HaJoSa that worked for me logging in on the Contacts desktop app.

I think this issue may be related to the latest High Sierra update? I am running it on 2 Macs with the same behavior. I just recently set up Radicale so I can't compare to previous OS X.

I am unable to write or sync contacts from the desktop Contacts or the iOS app although both will connect and login.

On the desktop Calendar I can read the server but not write to it, and it gets very beachbally when syncing to the Radicale server. The iOS Calendar and Reminders apps both work perfectly. All other CalDAV clients I've tried work perfectly (Fantastical, BusyCal, AgenDAV).

I tried Baikal as well and had the exact same behavior with these clients. Will try a different CardDAV client and see if that works...

**Comment by HaJoSa** [comment]
@oakbrad Yes I think, High Sierra messed up especially carddav. So searched for some other apps to overcome the problems with the build-in Calendar and Contacts.
Now I'm testing successfully **BusyCal** and **BusyContacts** on a MacBook with the latest High Sierra OS.
There are some problems left on an old iPad running IOS 9. iPhones with IOS 11 work perfectly with calendar and contacts.

**Comment by oakbrad** [comment]
You are right I tried BusyContacts and was able to import my address book into the server. Now Contacts can read that, but it won't write. The iOS app works perfectly.

Not sure what Apple did here to break things on the desktop, happy to provide logs or more info if devs can point me to where that would be.

**Comment by HaJoSa** [comment]
To end the never ending story I'll check if BusyCal and BusyContacts will be running on an iMac with HighSierra. If so my friend has to decide buying and using BusyCal & BusyContacts instead of the Apple apps.

**Comment by HaJoSa** [comment]
My friend bought BusyCal and BusyContacts and now his iMac and Macbook both sync fine with Radicale.
On his old iPad running IOS 9 he installed BusyCal for IOS and successfully syncronized with Radicale, too.
So BusyCal & BusyContacts may be an option for others who have problems with syncing to Radicale.

**Comment by Madeck** [comment]
@skyuplam I am still having this issue, and as I am using Radicale on a QNAP, I don't have access to the web server directly. I believe the option to do redirect well-known by default has been removed in #451 and discussed in #126, but at one point is was possible to control this via the config file.

I would be happy to know if there was another way to make macOS Contacts on High Sierra work. I have activated the debug mode in Radicale, but I am unsure how to get useful logs for the client Contacts app. Any help is is welcome.

**Comment by skyuplam** [comment]
@mxdvl I don't have a QNAP, but you can do the redirect/routing via any kind of routing capabled application to re-router the traffic, for example, your firewall, like `iptables`/`nftables`.
Perhaps, you can simply modify the source of `Radicale` to fit your needs.

**Comment by mxdvl** [comment]
I will try to figure out how to do that redirect, but looking at my current setup, it seems that actually, my contacts are syncing, albeit only partially. Only the first address book and its name is not converted into a group name. This seem to suggest that the issue has little to do with the /.well-know/carddav redirect.

![screen shot 2018-06-04 at 16 35 45](https://user-images.githubusercontent.com/76776/40904339-6a3daa8c-6815-11e8-8cc5-3dab32b01243.png)


Looking at the Radicale logs, I can see that the macOS AddressBook is making REPORT requests for the **/user/addressbook-1/** but never **/user/adressbook-2/**. When creating a group using the app, it creates a contact that can only be seen by the macOS app (above called "New Group of Contacts").

I would be more than happy to help debug this and file a bug report towards Apple if the problem lies within the macOS Contacts app.

**Comment by AddoSolutions** [comment]
I am having the same issue as rajid, identical setup

**Comment by rajid** [comment]
I finally moved to using "calendarserver" under Ubuntu. It seems to work just fine. Of course, you don't have Apple push notifications, but you can't have that with Radicale either. I'm actually going to just try to continue using Apple's server for contacts and calendar. Just move all other services (mail, web, etc.) to the new external server. This means I can still ultimately move to a residential connection, since contacts and calendar can be run on any port.

**Comment by royalhacker** [comment]
is the issue resolved ?

**Comment by rajid** [comment]
I decided to stick with Apple server for Contacts, at least for now, since it also supports push notifications, on which Apple's clients tend to depend (since they don't have a manual "update" function in most cases).

/raj

**Comment by royalhacker** [comment]
Contacts by default searches on .well-known/caldav 
`rewrite ^/.well-known/carddav /radicale/ redirect;`
`rewrite ^/.well-known/caldav /radicale/ redirect;`
Use this redirect for nginx

**Comment by Tntdruid** [comment]
I use 
_carddavs._tcp 86400 IN SRV 10 20 443 dav.example.org.
_caldavs._tcp 86400 IN SRV 10 20 443 dav.example.org.

**Comment by C4ne** [comment]
Note that although you can disable SSL in the iOS and MacOS client, Apple won't let you add that account over plain http. So you will have to create a (self signed) certificate to use tls encryption.
At least that's the case for my devices: MacOS Mojave and iOS 13.
In combination with @Tntdruid 's trick, that did it for me ... what a hassle.

**Comment by return42** [comment]
> .. what a hassle.

Its Apple, that are the reasons why I kicked out all my apple stuff years ago.

**Comment by tieugene** [comment]
Same problem. No CalDAV nor CardDAV work in macOS 10.15.7 (Catalina).
Tried to do as for Third (and similar to [nice documentation](http://return42.github.io/handsOn/radicale/dav_client.html#apple-kontakte)) - radicale (Fedora 33, Radicale 3.0, ru_RU.UTF8 locale) log filled with esoteric messages and nothing works:
```
Dec  6 23:45:42 localhost radicale[6190]: [2020-12-06 23:45:42 +0300] [6190/Thread-285] [ERROR] An error occurred during request: code 400, message Bad request version ('\x14G\x9c\x81O')
Dec  6 23:45:42 localhost radicale[6190]: [2020-12-06 23:45:42 +0300] [6190/Thread-286] [ERROR] An error occurred during request: code 400, message Bad request version ('_ö\x86\x9c5\x91r\x006\x13\x01\x13\x02\x13\x03À,À+À$À#À')
Dec  6 23:45:42 localhost radicale[6190]: [2020-12-06 23:45:42 +0300] [6190/Thread-287] [ERROR] An error occurred during request: code 400, message Bad request version ('(Õ\x07')
```

**Comment by bvrulez** [comment]
> I just got it to work on my newly installed radicale (on port 5253) and on my High Sierra MacOS. I used "Manual" configuration and gave the server address as "http://192.168.0.36:5253/raj/". (Yes, this is only inside my house for now.) After creating the addressbook using the radicale web interface, this configuration worked to access it on my mac.
> 

This does not seem to work any more.

**Comment by da4089** [comment]
> Same problem. No CalDAV nor CardDAV work in macOS 10.15.7 (Catalina). Tried to do as for Third (and similar to [nice documentation](http://return42.github.io/handsOn/radicale/dav_client.html#apple-kontakte)) - radicale (Fedora 33, Radicale 3.0, ru_RU.UTF8 locale) log filled with esoteric messages and nothing works:
> 
> ```
> Dec  6 23:45:42 localhost radicale[6190]: [2020-12-06 23:45:42 +0300] [6190/Thread-285] [ERROR] An error occurred during request: code 400, message Bad request version ('\x14G\x9c\x81O')
> Dec  6 23:45:42 localhost radicale[6190]: [2020-12-06 23:45:42 +0300] [6190/Thread-286] [ERROR] An error occurred during request: code 400, message Bad request version ('_ö\x86\x9c5\x91r\x006\x13\x01\x13\x02\x13\x03À,À+À$À#À')
> Dec  6 23:45:42 localhost radicale[6190]: [2020-12-06 23:45:42 +0300] [6190/Thread-287] [ERROR] An error occurred during request: code 400, message Bad request version ('(Õ\x07')
> ```

FWIW, I see a similar result.

macOS Big Sur, 11.7.10
Radicale 3.1.8 (from PyPI) running with --debug.
Using default macOS Contacts app

I used a "Manual" configuration, the same username/password as I used to load the VCF file.
Setting the "Server Address" to a URL showed nothing in the server logs.
Setting the "Server Address" to just "127.0.0.1:5232", gets me similar logs to @viertelb.
Running the server with --debug, I just got "Unable to verify account name or password" messages in the UI when configured

**Comment by pbiering** [comment]
closed now, please reopen if original topic still exists

**Comment by romuloalves** [comment]
> Contacts by default searches on .well-known/caldav `rewrite ^/.well-known/carddav /radicale/ redirect;` `rewrite ^/.well-known/caldav /radicale/ redirect;` Use this redirect for nginx

This fixed the problem for me.

**Comment by pbiering** [comment]
@romuloalves - can you please review https://github.com/Kozea/Radicale/pull/1696/files - thank you!

**Comment by pbiering** [comment]
please try latest examples for reverse proxy
